heartbeat/processor/peerAuthenticationRequestsProcessor.go

```diff
@@ -0,0 +1,234 @@
+	for {
+		if processor.isThresholdReached(sortedValidatorsKeys) {
+			log.Debug("received enough messages, closing PeerAuthenticationRequestsProcessor go routine")
+			return
+		}
+
+		requestsTimer.Reset(processor.delayBetweenRequests)
+		select {
+		case <-requestsTimer.C:
+			processor.requestMissingKeys(sortedValidatorsKeys)
+		case <-timeoutTimer.C:
+			log.Debug("timeout reached, not enough messages received, closing PeerAuthenticationRequestsProcessor go routine")
+			return
+		case <-ctx.Done():
+			log.Debug("closing PeerAuthenticationRequestsProcessor go routine")
+			return
+		}
+	}
+}
+
+func (processor *PeerAuthenticationRequestsProcessor) requestKeysChunks(keys [][]byte) {
+	maxChunks := processor.getMaxChunks(keys)
+	for chunkIndex := uint32(0); chunkIndex < maxChunks; chunkIndex++ {
+		processor.requestHandler.RequestPeerAuthenticationsChunk(processor.shardId, chunkIndex)
+
+		time.Sleep(processor.delayBetweenRequests)
+	}
+}
+
+func (processor *PeerAuthenticationRequestsProcessor) getSortedValidatorsKeys() ([][]byte, error) {
+	validatorsPKsMap, err := processor.nodesCoordinator.GetAllEligibleValidatorsPublicKeys(processor.epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	validatorsPKs := make([][]byte, 0)
+	for _, shardValidators := range validatorsPKsMap {
+		validatorsPKs = append(validatorsPKs, shardValidators...)
+	}
+
+	sort.Slice(validatorsPKs, func(i, j int) bool {
+		return bytes.Compare(validatorsPKs[i], validatorsPKs[j]) < 0
+	})
+
+	return validatorsPKs, nil
+}
+
+func (processor *PeerAuthenticationRequestsProcessor) getMaxChunks(dataBuff [][]byte) uint32 {
+	maxChunks := len(dataBuff) / int(processor.messagesInChunk)
+	if len(dataBuff)%int(processor.messagesInChunk) != 0 {
+		maxChunks++
+	}
+
+	return uint32(maxChunks)
+}
+
+func (processor *PeerAuthenticationRequestsProcessor) isThresholdReached(sortedValidatorsKeys [][]byte) bool {
+	minKeysExpected := float32(len(sortedValidatorsKeys)) * processor.minPeersThreshold
+	keysInCache := processor.peerAuthenticationPool.Keys()
+
+	return float32(len(keysInCache)) >= minKeysExpected
+}
+
+func (processor *PeerAuthenticationRequestsProcessor) requestMissingKeys(sortedValidatorsKeys [][]byte) {
+	missingKeys := processor.getMissingKeys(sortedValidatorsKeys)
+	if len(missingKeys) == 0 {
+		return
+	}
+
+	processor.requestHandler.RequestPeerAuthenticationsByHashes(processor.shardId, missingKeys)
+}
+
+func (processor *PeerAuthenticationRequestsProcessor) getMissingKeys(sortedValidatorsKeys [][]byte) [][]byte {
+	validatorsMap := make(map[string]bool, len(sortedValidatorsKeys))
+	for _, key := range sortedValidatorsKeys {
+		validatorsMap[string(key)] = false
+	}
+
+	keysInCache := processor.peerAuthenticationPool.Keys()
+	for _, key := range keysInCache {
+		validatorsMap[string(key)] = true
+	}
+
+	missingKeys := make([][]byte, 0)
+	for mKey, mVal := range validatorsMap {
+		if !mVal {
+			missingKeys = append(missingKeys, []byte(mKey))
+		}
+	}
+
+	return missingKeys
```

Comment:
this will work for about 10000 missing keys.
Before returning this, I would add a new function that "randomly" creates a sub-slice out of all missing keys if they exceed a certain number of items (lets say 1000 - can be configurable)